Logit link saturating

I've been running into `ERROR: AssertionError: isfinite(crit)` errors from https://github.com/JuliaStats/GLM.jl/blob/49853b1283da59183a537783c5296aa0d27b194e/src/glmfit.jl#L268 when trying to fit a GLM to Bernoulli response values and continuous predictors (~10,000 samples, each ~90 dimensional) with a logit link. The two distributions are quite well separated. 

The problem seems to originate here: https://github.com/JuliaStats/GLM.jl/blob/49853b1283da59183a537783c5296aa0d27b194e/src/glmfit.jl#L113
Which returns `dμdη = 0` for `η[i] = 753.8917511344454` and `L = LogitLink`, which then causes a `nan` by dividing by zero.

I've been seeing the same `AssertionError: isfinite(crit)` with `Binomial(), LogitLink()`

```
AssertionError: isfinite(crit)

Stacktrace:
 [1] _fit!(::GeneralizedLinearModel{GlmResp{Array{Float64,1},Binomial{Float64},LogitLink},DensePredChol{Float64,LinearAlgebra.Cholesky{Float64,Array{Float64,2}}}}, ::Bool, ::Int64, ::Float64, ::Float64, ::Nothing) at /Users/IanB/.julia/packages/GLM/YjBzj/src/glmfit.jl:268
```

Here's an example that produces the error I'm seeing:
```julia
x = rand(1000)
df = DataFrame(class=x.>0.5, a=x, b=rand(1000))
ols = glm(@formula(class ~ a + b), df, Binomial(),LogitLink())
```
Variable `a` is dominantly linked to `class`, while `b` introduces some noise, so glm should be able to resolve a model heavily linked to `a`

```
AssertionError: isfinite(crit)

Stacktrace:
 [1] _fit!(::GeneralizedLinearModel{GlmResp{Array{Float64,1},Binomial{Float64},LogitLink},DensePredChol{Float64,LinearAlgebra.Cholesky{Float64,Array{Float64,2}}}}, ::Bool, ::Int64, ::Float64, ::Float64, ::Nothing) at /Users/IanB/.julia/packages/GLM/YjBzj/src/glmfit.jl:268
 [2] #fit!#6 at /Users/IanB/.julia/packages/GLM/YjBzj/src/glmfit.jl:277 [inlined]
 [3] #fit! at ./none:0 [inlined]
 [4] #fit#8(::Bool, ::Array{Float64,1}, ::Array{Float64,1}, ::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol},NamedTuple{(:convTol,),Tuple{Float64}}}, ::typeof(fit), ::Type{GeneralizedLinearModel}, ::Array{Float64,2}, ::Array{Float64,1}, ::Binomial{Float64}, ::LogitLink) at /Users/IanB/.julia/packages/GLM/YjBzj/src/glmfit.jl:316
 [5] #fit at ./none:0 [inlined]
 [6] #fit#9(::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol},NamedTuple{(:convTol,),Tuple{Float64}}}, ::typeof(fit), ::Type{GeneralizedLinearModel}, ::Array{Float64,2}, ::Array{Bool,1}, ::Binomial{Float64}, ::LogitLink) at /Users/IanB/.julia/packages/GLM/YjBzj/src/glmfit.jl:323
 [7] (::getfield(StatsBase, Symbol("#kw##fit")))(::NamedTuple{(:convTol,),Tuple{Float64}}, ::typeof(fit), ::Type{GeneralizedLinearModel}, ::Array{Float64,2}, ::Array{Bool,1}, ::Binomial{Float64}, ::LogitLink) at ./none:0
 [8] #fit#36(::Dict{Any,Any}, ::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol},NamedTuple{(:convTol,),Tuple{Float64}}}, ::typeof(fit), ::Type{GeneralizedLinearModel}, ::Formula, ::DataFrame, ::Binomial{Float64}, ::Vararg{Any,N} where N) at /Users/IanB/.julia/packages/StatsModels/pBxdt/src/statsmodel.jl:72
 [9] (::getfield(StatsBase, Symbol("#kw##fit")))(::NamedTuple{(:convTol,),Tuple{Float64}}, ::typeof(fit), ::Type{GeneralizedLinearModel}, ::Formula, ::DataFrame, ::Binomial{Float64}, ::LogitLink) at ./none:0
 [10] #glm#10(::Base.Iterators.Pairs{Symbol,Float64,Tuple{Symbol},NamedTuple{(:convTol,),Tuple{Float64}}}, ::typeof(glm), ::Formula, ::DataFrame, ::Binomial{Float64}, ::Vararg{Any,N} where N) at /Users/IanB/.julia/packages/GLM/YjBzj/src/glmfit.jl:326
 [11] (::getfield(GLM, Symbol("#kw##glm")))(::NamedTuple{(:convTol,),Tuple{Float64}}, ::typeof(glm), ::Formula, ::DataFrame, ::Binomial{Float64}, ::Vararg{Any,N} where N) at ./none:0
 [12] top-level scope at In[44]:3
```

Interesting point.. It could theoretically be the case. Perhaps that's a situation that could be caught somehow and terminate early? 

It turned out the original reason I was hitting the `AssertionError: isfinite(crit)` was due to `Inf` values in my dataframe that I hadn't noticed. I guess most people would notice things like that, but a simple check and early fail for `Inf` in input data might be helpful

--- a/Project.toml
+++ b/Project.toml
@@ -28,7 +28,7 @@ test = ["CategoricalArrays", "CSV", "DataFrames", "RDatasets", "Test"]
 [compat]
 CategoricalArrays = "0.3, 0.4, 0.5"
 CSV = "0.2, 0.3, 0.4"
-DataFrames = "0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17"
+DataFrames = "0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18"
 Distributions = "0.16, 0.17, 0.18, 0.19"
 RDatasets = "0.5, 0.6"
 Reexport = "0.1, 0.2"

--- a/src/glmfit.jl
+++ b/src/glmfit.jl
@@ -355,14 +355,14 @@ function StatsBase.fit!(m::AbstractGLM;
     end
     if haskey(kwargs, :convTol)
         Base.depwarn("'convTol' argument is deprecated, use `atol` and `rtol` instead", :fit!)
-        tol = kwargs[:convTol]
+        rtol = kwargs[:convTol]
     end
     if !issubset(keys(kwargs), (:maxIter, :minStepFac, :convTol))
         throw(ArgumentError("unsupported keyword argument"))
     end
     if haskey(kwargs, :tol)
         Base.depwarn("`tol` argument is deprecated, use `atol` and `rtol` instead", :fit!)
-        tol = kwargs[:tol]
+        rtol = kwargs[:tol]
     end
 
     _fit!(m, verbose, maxiter, minstepfac, atol, rtol, start)
@@ -390,14 +390,14 @@ function StatsBase.fit!(m::AbstractGLM,
     end
     if haskey(kwargs, :convTol)
         Base.depwarn("'convTol' argument is deprecated, use `atol` and `rtol` instead", :fit!)
-        tol = kwargs[:convTol]
+        rtol = kwargs[:convTol]
     end
     if !issubset(keys(kwargs), (:maxIter, :minStepFac, :convTol))
         throw(ArgumentError("unsupported keyword argument"))
     end
     if haskey(kwargs, :tol)
         Base.depwarn("`tol` argument is deprecated, use `atol` and `rtol` instead", :fit!)
-        tol = kwargs[:tol]
+        rtol = kwargs[:tol]
     end
 
     r = m.rr

--- a/src/glmfit.jl
+++ b/src/glmfit.jl
@@ -106,16 +106,79 @@ function updateμ!(r::GlmResp{V,D,L}) where {V<:FPVector,D,L}
     end
 end
 
+function _weights_residuals(yᵢ, ηᵢ, μᵢ, omμᵢ, dμdηᵢ, l::LogitLink)
+    # LogitLink is the canonical link function for Binomial so only wrkresᵢ can
+    # possibly fail when dμdη==0 in which case it evaluates to ±1.
+    if iszero(dμdηᵢ)
+        wrkresᵢ = ifelse(yᵢ == 1, one(μᵢ), -one(μᵢ))
+    else
+        wrkresᵢ = ifelse(yᵢ == 1, omμᵢ, yᵢ - μᵢ) / dμdηᵢ
+    end
+    wrkwtᵢ = μᵢ*omμᵢ
+
+    return wrkresᵢ, wrkwtᵢ
+end
+
+function _weights_residuals(yᵢ, ηᵢ, μᵢ, omμᵢ, dμdηᵢ, l::ProbitLink)
+    # Since μomμ will underflow before dμdη for Probit, we can just check the
+    # former to decide when to evaluate with the tail approximation.
+    μomμᵢ = μᵢ*omμᵢ
+    if iszero(μomμᵢ)
+        wrkresᵢ = 1/abs(ηᵢ)
+        wrkwtᵢ  = dμdηᵢ
+    else
+        wrkresᵢ = ifelse(yᵢ == 1, omμᵢ, yᵢ - μᵢ) / dμdηᵢ
+        wrkwtᵢ  = abs2(dμdηᵢ)/μomμᵢ
+    end
+
+    return wrkresᵢ, wrkwtᵢ
+end
+
+function _weights_residuals(yᵢ, ηᵢ, μᵢ, omμᵢ, dμdηᵢ, l::CloglogLink)
+    if yᵢ == 1
+        wrkresᵢ = exp(-ηᵢ)
+    else
+        emη = exp(-ηᵢ)
+        if iszero(emη)
+            # Diverges to -∞
+            wrkresᵢ = -typeof(wrkresᵢ)(Inf)
+        elseif isinf(emη)
+            # converges to -1
+            wrkresᵢ = -one(emη)
+        else
+            wrkresᵢ = (yᵢ - μᵢ)/omμᵢ*emη
+        end
+    end
+
+    wrkwtᵢ  = exp(2*ηᵢ)/expm1(exp(ηᵢ))
+    # We know that both limits are zero so we'll convert NaNs
+    wrkwtᵢ = ifelse(isnan(wrkwtᵢ), zero(wrkwtᵢ), wrkwtᵢ)
+
+    return wrkresᵢ, wrkwtᵢ
+end
+
+# Fallback for remaining link functions
+function _weights_residuals(yᵢ, ηᵢ, μᵢ, omμᵢ, dμdηᵢ, l::Link01)
+    wrkresᵢ = ifelse(yᵢ == 1, omμᵢ, yᵢ - μᵢ)/dμdηᵢ
+    wrkwtᵢ  = abs2(dμdηᵢ)/(μᵢ*omμᵢ)
+
+    return wrkresᵢ, wrkwtᵢ
+end
+
 function updateμ!(r::GlmResp{V,D,L}) where {V<:FPVector,D<:Union{Bernoulli,Binomial},L<:Link01}
     y, η, μ, wrkres, wrkwt, dres = r.y, r.eta, r.mu, r.wrkresid, r.wrkwt, r.devresid
 
     @inbounds for i in eachindex(y, η, μ, wrkres, wrkwt, dres)
-        μi, dμdη, μomμ = inverselink(L(), η[i])
-        μ[i] = μi
-        yi = y[i]
-        wrkres[i] = (yi - μi) / dμdη
-        wrkwt[i] = cancancel(r) ? dμdη : abs2(dμdη) / μomμ
-        dres[i] = devresid(r.d, yi, μi)
+        yᵢ, ηᵢ = y[i], η[i]
+        μᵢ, omμᵢ, dμdηᵢ = inverselink(L(), ηᵢ)
+        μ[i] = μᵢ
+        # For large values of ηᵢ the quantities dμdη and μomμ will underflow.
+        # The ratios defining (yᵢ - μᵢ)/dμdη and dμdη^2/μomμ have fairly stable
+        # tail behavior so we can switch algorithm to avoid 0/0. The behavior
+        # is specific to the link function so _weights_residuals dispatches to
+        # robust versions for LogitLink and ProbitLink
+        wrkres[i], wrkwt[i] = _weights_residuals(yᵢ, ηᵢ, μᵢ, omμᵢ, dμdηᵢ, L())
+        dres[i] = devresid(r.d, yᵢ, μᵢ)
     end
 end
 
@@ -200,7 +263,7 @@ end
 dof(x::GeneralizedLinearModel) = dispersion_parameter(x.rr.d) ? length(coef(x)) + 1 : length(coef(x))
 
 function _fit!(m::AbstractGLM, verbose::Bool, maxiter::Integer, minstepfac::Real,
-               tol::Real, start)
+               atol::Real, rtol::Real, start)
 
     # Return early if model has the fit flag set
     m.fit && return m
@@ -246,9 +309,9 @@ function _fit!(m::AbstractGLM, verbose::Bool, maxiter::Integer, minstepfac::Real
 
         # Line search
         ## If the deviance isn't declining then half the step size
-        ## The tol*dev term is to avoid failure when deviance
+        ## The rtol*dev term is to avoid failure when deviance
         ## is unchanged except for rouding errors.
-        while dev > devold + tol*dev
+        while dev > devold + rtol*dev
             f /= 2
             f > minstepfac || error("step-halving failed at beta0 = $(p.beta0)")
             try
@@ -261,13 +324,12 @@ function _fit!(m::AbstractGLM, verbose::Bool, maxiter::Integer, minstepfac::Real
         installbeta!(p, f)
 
         # Test for convergence
-        crit = (devold - dev)/dev
-        verbose && println("$i: $dev, $crit")
-        if crit < tol || dev == 0
+        verbose && println("Iteration: $i, deviance: $dev, diff.dev.:$(devold - dev)")
+        if devold - dev < max(rtol*devold, atol)
             cvg = true
             break
         end
-        @assert isfinite(crit)
+        @assert isfinite(dev)
         devold = dev
     end
     cvg || throw(ConvergenceException(maxiter))
@@ -275,8 +337,13 @@ function _fit!(m::AbstractGLM, verbose::Bool, maxiter::Integer, minstepfac::Real
     m
 end
 
-function StatsBase.fit!(m::AbstractGLM; verbose::Bool=false, maxiter::Integer=30,
-                        minstepfac::Real=0.001, tol::Real=1e-6, start=nothing,
+function StatsBase.fit!(m::AbstractGLM;
+                        verbose::Bool=false,
+                        maxiter::Integer=30,
+                        minstepfac::Real=0.001,
+                        atol::Real=1e-6,
+                        rtol::Real=1e-6,
+                        start=nothing,
                         kwargs...)
     if haskey(kwargs, :maxIter)
         Base.depwarn("'maxIter' argument is deprecated, use 'maxiter' instead", :fit!)
@@ -287,19 +354,32 @@ function StatsBase.fit!(m::AbstractGLM; verbose::Bool=false, maxiter::Integer=30
         minstepfac = kwargs[:minStepFac]
     end
     if haskey(kwargs, :convTol)
-        Base.depwarn("'convTol' argument is deprecated, use 'tol' instead", :fit!)
+        Base.depwarn("'convTol' argument is deprecated, use `atol` and `rtol` instead", :fit!)
         tol = kwargs[:convTol]
     end
     if !issubset(keys(kwargs), (:maxIter, :minStepFac, :convTol))
         throw(ArgumentError("unsupported keyword argument"))
     end
+    if haskey(kwargs, :tol)
+        Base.depwarn("`tol` argument is deprecated, use `atol` and `rtol` instead", :fit!)
+        tol = kwargs[:tol]
+    end
 
-    _fit!(m, verbose, maxiter, minstepfac, tol, start)
+    _fit!(m, verbose, maxiter, minstepfac, atol, rtol, start)
 end
 
-function StatsBase.fit!(m::AbstractGLM, y; wts=nothing, offset=nothing, dofit::Bool=true,
-                        verbose::Bool=false, maxiter::Integer=30, minstepfac::Real=0.001,
-                        tol::Real=1e-6, start=nothing, kwargs...)
+function StatsBase.fit!(m::AbstractGLM,
+                        y;
+                        wts=nothing,
+                        offset=nothing,
+                        dofit::Bool=true,
+                        verbose::Bool=false,
+                        maxiter::Integer=30,
+                        minstepfac::Real=0.001,
+                        atol::Real=1e-6,
+                        rtol::Real=1e-6,
+                        start=nothing,
+                        kwargs...)
     if haskey(kwargs, :maxIter)
         Base.depwarn("'maxIter' argument is deprecated, use 'maxiter' instead", :fit!)
         maxiter = kwargs[:maxIter]
@@ -309,11 +389,15 @@ function StatsBase.fit!(m::AbstractGLM, y; wts=nothing, offset=nothing, dofit::B
         minstepfac = kwargs[:minStepFac]
     end
     if haskey(kwargs, :convTol)
-        Base.depwarn("'convTol' argument is deprecated, use 'tol' instead", :fit!)
+        Base.depwarn("'convTol' argument is deprecated, use `atol` and `rtol` instead", :fit!)
         tol = kwargs[:convTol]
     end
     if !issubset(keys(kwargs), (:maxIter, :minStepFac, :convTol))
         throw(ArgumentError("unsupported keyword argument"))
+    end
+    if haskey(kwargs, :tol)
+        Base.depwarn("`tol` argument is deprecated, use `atol` and `rtol` instead", :fit!)
+        tol = kwargs[:tol]
     end
 
     r = m.rr
@@ -326,7 +410,7 @@ function StatsBase.fit!(m::AbstractGLM, y; wts=nothing, offset=nothing, dofit::B
     fill!(m.pp.beta0, 0)
     m.fit = false
     if dofit
-        _fit!(m, verbose, maxiter, minstepfac, tol, start)
+        _fit!(m, verbose, maxiter, minstepfac, atol, rtol, start)
     else
         m
     end
@@ -346,8 +430,10 @@ vector, respectively, or a formula and a data frame. `d` must be a
 length 0
 - `verbose::Bool=false`: Display convergence information for each iteration
 - `maxiter::Integer=30`: Maximum number of iterations allowed to achieve convergence
-- `tol::Real=1e-6`: Convergence is achieved when the relative change in
-deviance is less than this
+- `atol::Real=1e-6`: Convergence is achieved when the relative change in
+deviance is less than `max(rtol*dev, atol)`.
+- `rtol::Real=1e-6`: Convergence is achieved when the relative change in
+deviance is less than `max(rtol*dev, atol)`.
 - `minstepfac::Real=0.001`: Minimum line step fraction. Must be between 0 and 1.
 - `start::AbstractVector=nothing`: Starting values for beta. Should have the
 same length as the number of columns in the model matrix.
@@ -373,11 +459,11 @@ function fit(::Type{M},
 end
 
 fit(::Type{M},
-X::Union{Matrix,SparseMatrixCSC},
-y::AbstractVector,
-d::UnivariateDistribution,
-l::Link=canonicallink(d); kwargs...) where {M<:AbstractGLM} =
-    fit(M, float(X), float(y), d, l; kwargs...)
+    X::Union{Matrix,SparseMatrixCSC},
+    y::AbstractVector,
+    d::UnivariateDistribution,
+    l::Link=canonicallink(d); kwargs...) where {M<:AbstractGLM} =
+        fit(M, float(X), float(y), d, l; kwargs...)
 
 """
     glm(F, D, args...; kwargs...)
@@ -485,4 +571,3 @@ function checky(y, d::Binomial)
     end
     return nothing
 end
-

--- a/src/glmtools.jl
+++ b/src/glmtools.jl
@@ -95,7 +95,7 @@ A [`Link`](@ref) defined as `η = √μ`
 mutable struct SqrtLink <: Link end
 
 """
-    linkfun(L::Link, μ)
+    linkfun(L::Link, μ::Real)
 
 Return `η`, the value of the linear predictor for link `L` at mean `μ`.
 
@@ -112,7 +112,7 @@ julia> show(linkfun.(LogitLink(), μ))
 function linkfun end
 
 """
-    linkinv(L::Link, η)
+    linkinv(L::Link, η::Real)
 
 Return `μ`, the mean value, for link `L` at linear predictor value `η`.
 
@@ -123,14 +123,14 @@ julia> μ = 0.1:0.2:1
 
 julia> η = logit.(μ);
 
-julia> linkinv.(LogitLink(), η) ≈ μ
+julia> linkinv.(LogitLink(), η::Real) ≈ μ
 true
 ```
 """
 function linkinv end
 
 """
-    mueta(L::Link, η)
+    mueta(L::Link, η::Real)
 
 Return the derivative of [`linkinv`](@ref), `dμ/dη`, for link `L` at linear predictor value `η`.
 
@@ -149,7 +149,7 @@ true
 function mueta end
 
 """
-    inverselink(L::Link, η)
+    inverselink(L::Link, η::Real)
 
 Return a 3-tuple of the inverse link, the derivative of the inverse link, and when appropriate, the variance function `μ*(1 - μ)`.
 
@@ -187,94 +187,96 @@ LogitLink()
 """
 function canonicallink end
 
-linkfun(::CauchitLink, μ) = tan(pi * (μ - oftype(μ, 1/2)))
-linkinv(::CauchitLink, η) = oftype(η, 1/2) + atan(η) / pi
-mueta(::CauchitLink, η) = one(η) / (pi * (one(η) + abs2(η)))
-function inverselink(::CauchitLink, η)
-    μlower = atan(-abs(η)) / π
-    μlower += oftype(μlower, 1/2)
-    η > 0 ? 1 - μlower : μlower, inv(π * (1 + abs2(η))), μlower * (1 - μlower)
+linkfun(::CauchitLink, μ::Real) = tan(pi * (μ - oftype(μ, 1/2)))
+linkinv(::CauchitLink, η::Real) = oftype(η, 1/2) + atan(η) / pi
+mueta(::CauchitLink, η::Real) = one(η) / (pi * (one(η) + abs2(η)))
+function inverselink(::CauchitLink, η::Real)
+    # atan decays so slowly that we don't need to be careful when evaluating μ
+    μ = atan(η) / π
+    μ += one(μ)/2
+    return μ, 1 - μ, inv(π * (1 + abs2(η)))
 end
 
-linkfun(::CloglogLink, μ) = log(-log1p(-μ))
-function linkinv(::CloglogLink, η::T) where T<:Real
-    clamp(-expm1(-exp(η)), eps(T), one(T) - eps(T))
-end
-function mueta(::CloglogLink, η::T) where T<:Real
-    max(eps(T), exp(η) * exp(-exp(η)))
-end
-function inverselink(::CloglogLink, η)
+linkfun(::CloglogLink, μ::Real) = log(-log1p(-μ))
+linkinv(::CloglogLink, η::Real) = -expm1(-exp(η))
+mueta(::CloglogLink, η::Real) = exp(η) * exp(-exp(η))
+function inverselink(::CloglogLink, η::Real)
     expη = exp(η)
     μ = -expm1(-expη)
     omμ = exp(-expη)   # the complement, 1 - μ
-    μ, max(floatmin(μ), expη * omμ), max(floatmin(μ), μ * omμ)
+    return μ, omμ, expη * omμ
 end
 
-linkfun(::IdentityLink, μ) = μ
-linkinv(::IdentityLink, η) = η
-mueta(::IdentityLink, η) = one(η)
-inverselink(::IdentityLink, η) = η, one(η), oftype(η, NaN)
+linkfun(::IdentityLink, μ::Real) = μ
+linkinv(::IdentityLink, η::Real) = η
+mueta(::IdentityLink, η::Real) = one(η)
+inverselink(::IdentityLink, η::Real) = η, one(η), oftype(η, NaN)
 
-linkfun(::InverseLink, μ) = inv(μ)
-linkinv(::InverseLink, η) = inv(η)
-mueta(::InverseLink, η) = -inv(abs2(η))
-function inverselink(::InverseLink, η)
+linkfun(::InverseLink, μ::Real) = inv(μ)
+linkinv(::InverseLink, η::Real) = inv(η)
+mueta(::InverseLink, η::Real) = -inv(abs2(η))
+function inverselink(::InverseLink, η::Real)
     μ = inv(η)
-    μ, -abs2(μ), oftype(μ, NaN)
+    return μ, -abs2(μ), oftype(μ, NaN)
 end
 
-linkfun(::InverseSquareLink, μ) = inv(abs2(μ))
-linkinv(::InverseSquareLink, η) = inv(sqrt(η))
-mueta(::InverseSquareLink, η) = -inv(2η*sqrt(η))
-function inverselink(::InverseSquareLink, η)
+linkfun(::InverseSquareLink, μ::Real) = inv(abs2(μ))
+linkinv(::InverseSquareLink, η::Real) = inv(sqrt(η))
+mueta(::InverseSquareLink, η::Real) = -inv(2η*sqrt(η))
+function inverselink(::InverseSquareLink, η::Real)
     μ = inv(sqrt(η))
-    μ, -μ / (2η), oftype(μ, NaN)
+    return μ, -μ / (2η), oftype(μ, NaN)
 end
 
-linkfun(::LogitLink, μ) = logit(μ)
-linkinv(::LogitLink, η) = logistic(η)
-function mueta(::LogitLink, η)
+linkfun(::LogitLink, μ::Real) = logit(μ)
+linkinv(::LogitLink, η::Real) = logistic(η)
+function mueta(::LogitLink, η::Real)
     expabs = exp(-abs(η))
     denom = 1 + expabs
-    (expabs / denom) / denom
+    return (expabs / denom) / denom
 end
-function inverselink(::LogitLink, η)
+function inverselink(::LogitLink, η::Real)
     expabs = exp(-abs(η))
     opexpabs = 1 + expabs
     deriv = (expabs / opexpabs) / opexpabs
-    η ≤ 0 ? expabs / opexpabs : inv(opexpabs), deriv, deriv
+    if η < 0
+        μ, omμ = expabs / opexpabs, 1 / opexpabs
+    else
+        μ, omμ = 1 / opexpabs, expabs / opexpabs
+    end
+    return μ, omμ, deriv
 end
 
-linkfun(::LogLink, μ) = log(μ)
-linkinv(::LogLink, η) = exp(η)
-mueta(::LogLink, η) = exp(η)
-function inverselink(::LogLink, η)
+linkfun(::LogLink, μ::Real) = log(μ)
+linkinv(::LogLink, η::Real) = exp(η)
+mueta(::LogLink, η::Real) = exp(η)
+function inverselink(::LogLink, η::Real)
     μ = exp(η)
-    μ, μ, oftype(μ, NaN)
+    return μ, μ, oftype(μ, NaN)
 end
 
-linkfun(nbl::NegativeBinomialLink, μ) = log(μ / (μ + nbl.θ))
-linkinv(nbl::NegativeBinomialLink, η) = ℯ^η * nbl.θ / (1-ℯ^η)
-mueta(nbl::NegativeBinomialLink, η) = ℯ^η * nbl.θ / (1-ℯ^η)
-function inverselink(nbl::NegativeBinomialLink, η)
-    μ = ℯ^η * nbl.θ / (1-ℯ^η)
+linkfun(nbl::NegativeBinomialLink, μ::Real) = log(μ / (μ + nbl.θ))
+linkinv(nbl::NegativeBinomialLink, η::Real) = -exp(η) * nbl.θ / expm1(η)
+mueta(nbl::NegativeBinomialLink, η::Real) = -exp(η) * nbl.θ / expm1(η)
+function inverselink(nbl::NegativeBinomialLink, η::Real)
+    μ = -exp(η) * nbl.θ / expm1(η)
     deriv = μ * (1 + μ / nbl.θ)
-    μ, deriv, oftype(μ, NaN)
+    return μ, deriv, oftype(μ, NaN)
 end
 
-linkfun(::ProbitLink, μ) = -sqrt2 * erfcinv(2μ)
-linkinv(::ProbitLink, η) = erfc(-η / sqrt2) / 2
-mueta(::ProbitLink, η) = exp(-abs2(η) / 2) / sqrt2π
-function inverselink(::ProbitLink, η)
-    μlower = erfc(abs(η) / sqrt2) / 2
-    μupper = 1 - μlower
-    η < 0 ? μlower : μupper, exp(-abs2(η) / 2 ) / sqrt2π, μlower * μupper
+linkfun(::ProbitLink, μ::Real) = -sqrt2 * erfcinv(2μ)
+linkinv(::ProbitLink, η::Real) = erfc(-η / sqrt2) / 2
+mueta(::ProbitLink, η::Real) = exp(-abs2(η) / 2) / sqrt2π
+function inverselink(::ProbitLink, η::Real)
+    μ   =  cdf(Normal(), η)
+    omμ = ccdf(Normal(), η)
+    return μ, omμ, pdf(Normal(), η)
 end
 
-linkfun(::SqrtLink, μ) = sqrt(μ)
-linkinv(::SqrtLink, η) = abs2(η)
-mueta(::SqrtLink, η) = 2η
-inverselink(::SqrtLink, η) = abs2(η), 2η, oftype(η, NaN)
+linkfun(::SqrtLink, μ::Real) = sqrt(μ)
+linkinv(::SqrtLink, η::Real) = abs2(η)
+mueta(::SqrtLink, η::Real) = 2η
+inverselink(::SqrtLink, η::Real) = abs2(η), 2η, oftype(η, NaN)
 
 canonicallink(::Bernoulli) = LogitLink()
 canonicallink(::Binomial) = LogitLink()
@@ -285,7 +287,7 @@ canonicallink(::Normal) = IdentityLink()
 canonicallink(::Poisson) = LogLink()
 
 """
-    glmvar(D::Distribution, μ)
+    glmvar(D::Distribution, μ::Real)
 
 Return the value of the variance function for `D` at `μ`
 
@@ -297,27 +299,27 @@ variance that depends on `μ`.
 ```jldoctest
 julia> μ = 1/6:1/3:1;
 
-julia> glmvar.(Normal(), μ)    # constant for Normal()
+julia> glmvar.(Normal(), μ::Real)    # constant for Normal()
 3-element Array{Float64,1}:
  1.0
  1.0
  1.0
 
-julia> glmvar.(Bernoulli(), μ) ≈ μ .* (1 .- μ)
+julia> glmvar.(Bernoulli(), μ::Real) ≈ μ .* (1 .- μ)
 true
 
-julia> glmvar.(Poisson(), μ) == μ
+julia> glmvar.(Poisson(), μ::Real) == μ
 true
 ```
 """
 function glmvar end
 
-glmvar(::Union{Bernoulli,Binomial}, μ) = μ * (1 - μ)
-glmvar(::Gamma, μ) = abs2(μ)
-glmvar(::InverseGaussian, μ) = μ^3
-glmvar(d::NegativeBinomial, μ) = μ * (1 + μ/d.r)
-glmvar(::Normal, μ) = one(μ)
-glmvar(::Poisson, μ) = μ
+glmvar(::Union{Bernoulli,Binomial}, μ::Real) = μ * (1 - μ)
+glmvar(::Gamma, μ::Real) = abs2(μ)
+glmvar(::InverseGaussian, μ::Real) = μ^3
+glmvar(d::NegativeBinomial, μ::Real) = μ * (1 + μ/d.r)
+glmvar(::Normal, μ::Real) = one(μ)
+glmvar(::Poisson, μ::Real) = μ
 
 """
     mustart(D::Distribution, y, wt)
@@ -362,7 +364,7 @@ function mustart(::Poisson, y, wt)
 end
 
 """
-    devresid(D, y, μ)
+    devresid(D, y, μ::Real)
 
 Return the squared deviance residual of `μ` from `y` for distribution `D`
 
@@ -387,7 +389,7 @@ true
 """
 function devresid end
 
-function devresid(::Bernoulli, y, μ)
+function devresid(::Bernoulli, y, μ::Real)
     if y == 1
         return -2 * log(μ)
     elseif y == 0
@@ -395,7 +397,7 @@ function devresid(::Bernoulli, y, μ)
     end
     throw(ArgumentError("y should be 0 or 1 (got $y)"))
 end
-function devresid(::Binomial, y, μ)
+function devresid(::Binomial, y, μ::Real)
     if y == 1
         return -2 * log(μ)
     elseif y == 0
@@ -404,15 +406,15 @@ function devresid(::Binomial, y, μ)
         return 2 * (y * (log(y) - log(μ)) + (1 - y)*(log1p(-y) - log1p(-μ)))
     end
 end
-devresid(::Gamma, y, μ) = -2 * (log(y / μ) - (y - μ) / μ)
-devresid(::InverseGaussian, y, μ) = abs2(y - μ) / (y * abs2(μ))
-function devresid(d::NegativeBinomial, y, μ)
+devresid(::Gamma, y, μ::Real) = -2 * (log(y / μ) - (y - μ) / μ)
+devresid(::InverseGaussian, y, μ::Real) = abs2(y - μ) / (y * abs2(μ))
+function devresid(d::NegativeBinomial, y, μ::Real)
     θ = d.r
     v = 2 * (xlogy(y, y / μ) + xlogy(y + θ, (μ + θ)/(y + θ)))
     return μ == 0 ? oftype(v, NaN) : v
 end
-devresid(::Normal, y, μ) = abs2(y - μ)
-devresid(::Poisson, y, μ) = 2 * (xlogy(y, y / μ) - (y - μ))
+devresid(::Normal, y, μ::Real) = abs2(y - μ)
+devresid(::Poisson, y, μ::Real) = 2 * (xlogy(y, y / μ) - (y - μ))
 
 """
     dispersion_parameter(D)  # not exported

--- a/src/linpred.jl
+++ b/src/linpred.jl
@@ -151,7 +151,8 @@ end
 function delbeta!(p::DensePredChol{T,<:Cholesky}, r::Vector{T}, wt::Vector{T}) where T<:BlasReal
     scr = mul!(p.scratchm1, Diagonal(wt), p.X)
     cholesky!(Hermitian(mul!(cholfactors(p.chol), transpose(scr), p.X), :U))
-    ldiv!(p.chol, mul!(p.delbeta, transpose(scr), r))
+    mul!(p.delbeta, transpose(scr), r)
+    ldiv!(p.chol, p.delbeta)
     p
 end
 

--- a/src/negbinfit.jl
+++ b/src/negbinfit.jl
@@ -52,15 +52,15 @@ function negbin(F,
         minstepfac = kwargs[:minStepFac]
     end
     if haskey(kwargs, :convTol)
-        Base.depwarn("'convTol' argument is deprecated, use 'tol' instead", :negbin)
-        tol = kwargs[:convTol]
+        Base.depwarn("`convTol` argument is deprecated, use `atol` and `rtol` instead", :negbin)
+        rtol = kwargs[:convTol]
     end
     if !issubset(keys(kwargs), (:maxIter, :minStepFac, :convTol))
         throw(ArgumentError("unsupported keyword argument"))
     end
     if haskey(kwargs, :tol)
         Base.depwarn("`tol` argument is deprecated, use `atol` and `rtol` instead", :negbin)
-        tol = kwargs[:tol]
+        rtol = kwargs[:tol]
     end
 
     maxiter >= 1 || throw(ArgumentError("maxiter must be positive"))

--- a/src/negbinfit.jl
+++ b/src/negbinfit.jl
@@ -13,7 +13,7 @@ function mle_for_θ(y::AbstractVector, μ::AbstractVector, wts::AbstractVector;
 
     unit_weights = length(wts) == 0
     if unit_weights
-        n = length(y) 
+        n = length(y)
         θ = n / sum((yi/μi - 1)^2 for (yi, μi) in zip(y, μ))
     else
         n = sum(wts)
@@ -34,36 +34,47 @@ function mle_for_θ(y::AbstractVector, μ::AbstractVector, wts::AbstractVector;
     θ
 end
 
-function negbin(F, D, args...;
-                initialθ::Real=Inf, maxiter::Integer=30, tol::Real=1.e-6,
-                verbose::Bool=false, kwargs...)
+function negbin(F,
+                D,
+                args...;
+                initialθ::Real=Inf,
+                maxiter::Integer=30,
+                atol::Real=1e-6,
+                rtol::Real=1.e-6,
+                verbose::Bool=false,
+                kwargs...)
     if haskey(kwargs, :maxIter)
-        Base.depwarn("'maxIter' argument is deprecated, use 'maxiter' instead", :fit!)
+        Base.depwarn("'maxIter' argument is deprecated, use 'maxiter' instead", :negbin)
         maxiter = kwargs[:maxIter]
     end
     if haskey(kwargs, :minStepFac)
-        Base.depwarn("'minStepFac' argument is deprecated, use 'minstepfac' instead", :fit!)
+        Base.depwarn("'minStepFac' argument is deprecated, use 'minstepfac' instead", :negbin)
         minstepfac = kwargs[:minStepFac]
     end
     if haskey(kwargs, :convTol)
-        Base.depwarn("'convTol' argument is deprecated, use 'tol' instead", :fit!)
+        Base.depwarn("'convTol' argument is deprecated, use 'tol' instead", :negbin)
         tol = kwargs[:convTol]
     end
     if !issubset(keys(kwargs), (:maxIter, :minStepFac, :convTol))
         throw(ArgumentError("unsupported keyword argument"))
     end
+    if haskey(kwargs, :tol)
+        Base.depwarn("`tol` argument is deprecated, use `atol` and `rtol` instead", :negbin)
+        tol = kwargs[:tol]
+    end
 
     maxiter >= 1 || throw(ArgumentError("maxiter must be positive"))
-    tol > 0  || throw(ArgumentError("tol must be positive"))
+    atol > 0  || throw(ArgumentError("atol must be positive"))
+    rtol > 0  || throw(ArgumentError("rtol must be positive"))
     initialθ > 0 || throw(ArgumentError("initialθ must be positive"))
 
     # fit a Poisson regression model if the user does not specify an initial θ
-    if isinf(initialθ) 
+    if isinf(initialθ)
         regmodel = glm(F, D, Poisson(), args...;
-                       maxiter=maxiter, tol=tol, verbose=verbose, kwargs...)
+                       maxiter=maxiter, atol=atol, rtol=rtol, verbose=verbose, kwargs...)
     else
         regmodel = glm(F, D, NegativeBinomial(initialθ), args...;
-                       maxiter=maxiter, tol=tol, verbose=verbose, kwargs...)
+                       maxiter=maxiter, atol=atol, rtol=rtol, verbose=verbose, kwargs...)
     end
 
     μ = regmodel.model.rr.mu
@@ -82,16 +93,16 @@ function negbin(F, D, args...;
 
     converged = false
     for i = 1:maxiter
-        if abs(ll0 - ll)/d + abs(δ) <= tol
+        if abs(ll0 - ll)/d + abs(δ) <= rtol
             converged = true
             break
         end
         verbose && println("[ Alternating iteration ", i, ", θ = ", θ, " ]")
         regmodel = glm(F, D, NegativeBinomial(θ), args...;
-                       maxiter=maxiter, tol=tol, verbose=verbose, kwargs...)
+                       maxiter=maxiter, atol=atol, rtol=rtol, verbose=verbose, kwargs...)
         μ = regmodel.model.rr.mu
         prevθ = θ
-        θ = mle_for_θ(y, μ, wts; maxiter=maxiter, tol=tol)
+        θ = mle_for_θ(y, μ, wts; maxiter=maxiter, tol=rtol)
         δ = prevθ - θ
         ll0 = ll
         ll = loglikelihood(regmodel)

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -92,22 +92,19 @@ end
 admit = CSV.read(joinpath(glm_datadir, "admit.csv"))
 admit[:rank] = categorical(admit[:rank])
 
-@testset "Binomial, Bernoulli, LogitLink" begin
-    for distr in (Binomial, Bernoulli)
-        gm2 = fit(GeneralizedLinearModel, @formula(admit ~ 1 + gre + gpa + rank), admit,
-                  distr())
-        @test GLM.cancancel(gm2.model.rr)
-        test_show(gm2)
-        @test dof(gm2) == 6
-        @test isapprox(deviance(gm2), 458.5174924758994)
-        @test isapprox(loglikelihood(gm2), -229.25874623794968)
-        @test isapprox(aic(gm2), 470.51749247589936)
-        @test isapprox(aicc(gm2), 470.7312329339146)
-        @test isapprox(bic(gm2), 494.4662797585473)
-        @test isapprox(coef(gm2),
-            [-3.9899786606380756, 0.0022644256521549004, 0.804037453515578,
-            -0.6754428594116578, -1.340203811748108, -1.5514636444657495])
-    end
+@testset "$distr with LogitLink" for distr in (Binomial, Bernoulli)
+    gm2 = fit(GeneralizedLinearModel, @formula(admit ~ 1 + gre + gpa + rank), admit, distr())
+    @test GLM.cancancel(gm2.model.rr)
+    test_show(gm2)
+    @test dof(gm2) == 6
+    @test deviance(gm2) ≈ 458.5174924758994
+    @test loglikelihood(gm2) ≈ -229.25874623794968
+    @test isapprox(aic(gm2), 470.51749247589936)
+    @test isapprox(aicc(gm2), 470.7312329339146)
+    @test isapprox(bic(gm2), 494.4662797585473)
+    @test isapprox(coef(gm2),
+        [-3.9899786606380756, 0.0022644256521549004, 0.804037453515578,
+         -0.6754428594116578, -1.340203811748108, -1.5514636444657495])
 end
 
 @testset "Bernoulli ProbitLink" begin
@@ -187,7 +184,7 @@ end
 
 @testset "Normal LogLink offset" begin
     gm7 = fit(GeneralizedLinearModel, @formula(Postwt ~ 1 + Prewt + Treat), anorexia,
-              Normal(), LogLink(), offset=Array{Float64}(anorexia[:Prewt]), tol=1e-8)
+              Normal(), LogLink(), offset=Array{Float64}(anorexia[:Prewt]), rtol=1e-8)
     @test !GLM.cancancel(gm7.model.rr)
     test_show(gm7)
     @test isapprox(deviance(gm7), 3265.207242977156)
@@ -237,23 +234,23 @@ end
 
 @testset "Gamma LogLink" begin
     gm9 = fit(GeneralizedLinearModel, @formula(lot1 ~ 1 + u), clotting, Gamma(), LogLink(),
-              tol=1e-8)
+              rtol=1e-8, atol=0.0)
     @test !GLM.cancancel(gm9.model.rr)
     test_show(gm9)
     @test dof(gm9) == 3
-    @test isapprox(deviance(gm9), 0.16260829451739)
-    @test isapprox(loglikelihood(gm9), -26.24082810384911)
-    @test isapprox(aic(gm9), 58.48165620769822)
-    @test isapprox(aicc(gm9), 63.28165620769822)
-    @test isapprox(bic(gm9), 59.07332993970688)
-    @test isapprox(coef(gm9), [5.50322528458221, -0.60191617825971])
-    @test isapprox(GLM.dispersion(gm9.model, true), 0.02435442293561081)
-    @test isapprox(stderror(gm9), [0.19030107482720, 0.05530784660144])
+    @test deviance(gm9) ≈ 0.16260829451739
+    @test loglikelihood(gm9) ≈ -26.24082810384911
+    @test aic(gm9) ≈ 58.48165620769822
+    @test aicc(gm9) ≈ 63.28165620769822
+    @test bic(gm9) ≈ 59.07332993970688
+    @test coef(gm9) ≈ [5.50322528458221, -0.60191617825971]
+    @test GLM.dispersion(gm9.model, true) ≈ 0.02435442293561081
+    @test stderror(gm9) ≈ [0.19030107482720, 0.05530784660144]
 end
 
 @testset "Gamma IdentityLink" begin
     gm10 = fit(GeneralizedLinearModel, @formula(lot1 ~ 1 + u), clotting, Gamma(), IdentityLink(),
-               tol=1e-8)
+               rtol=1e-8, atol=0.0)
     @test !GLM.cancancel(gm10.model.rr)
     test_show(gm10)
     @test dof(gm10) == 3
@@ -300,14 +297,12 @@ admit_agr2[:p] = admit_agr2[:admit] ./ admit_agr2[:count]
     test_show(gm15)
     @test dof(gm15) == 4
     @test nobs(gm15) == 400
-# The model matrix is singular so the deviance is essentially round-off error
-#    @test isapprox(deviance(gm15), -2.4424906541753456e-15, rtol = 1e-7)
-    @test isapprox(loglikelihood(gm15), -9.50254433604239)
-    @test isapprox(aic(gm15), 27.00508867208478)
-    @test isapprox(aicc(gm15), 27.106354494869592)
-    @test isapprox(bic(gm15), 42.970946860516705)
-    @test isapprox(coef(gm15),
-        [0.1643030512912767, -0.7500299832303851, -1.3646980342693287, -1.6867295867357475])
+    @test deviance(gm15) ≈ -2.4424906541753456e-15 atol = 1e-13
+    @test loglikelihood(gm15) ≈ -9.50254433604239
+    @test aic(gm15) ≈ 27.00508867208478
+    @test aicc(gm15) ≈ 27.106354494869592
+    @test bic(gm15) ≈ 42.970946860516705
+    @test coef(gm15) ≈ [0.1643030512912767, -0.7500299832303851, -1.3646980342693287, -1.6867295867357475]
 end
 
 # Weighted Gamma example (weights are totally made up)
@@ -620,3 +615,12 @@ end
     @test dof_residual(model1) == dof_residual(model2)
     @test dof_residual(model3) == dof_residual(model4)
 end
+
+@testset "Issue #286 (separable data)" begin
+    x  = rand(1000)
+    df = DataFrame(y = x .> 0.5, x₁ = x, x₂ = rand(1000))
+    @testset "Binomial with $l" for l in (LogitLink(), ProbitLink(), CauchitLink(), CloglogLink())
+        @test glm(@formula(y ~ x₁ + x₂), df, Binomial(), l, maxiter=40) isa StatsModels.DataFrameRegressionModel
+    end
+end
+

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -620,7 +620,7 @@ end
     x  = rand(1000)
     df = DataFrame(y = x .> 0.5, x₁ = x, x₂ = rand(1000))
     @testset "Binomial with $l" for l in (LogitLink(), ProbitLink(), CauchitLink(), CloglogLink())
-        @test glm(@formula(y ~ x₁ + x₂), df, Binomial(), l, maxiter=40) isa StatsModels.DataFrameRegressionModel
+        @test deviance(glm(@formula(y ~ x₁ + x₂), df, Binomial(), l, maxiter=40)) < 1e-6
     end
 end
 

